### DISCUSSION TITLE
Accessing previous state value as well as the current

### DISCUSSION THREAD

**Discussion by viveleroi** [pmndrs/zustand#2590] [Q&A]
Is it possible to access to the previous state value as well as the current?

We're using the below store to keep track of a selected node from a tree navigation component. When this state changes, any React component rerenders properly. However, we've encountered a scenario where we need to run additional logic not when the value changes, but when the value changes and *specific values* do not match the prior state. 

We have hooks to do this in react, like usePreviousState from react-use library, but I'm hopping there's a way in zustand to query the last value as well as the new.

```tsx
interface TreeNodeState {
  selectedNode: Maybe<TreeNode>
}

const initialState: TreeNodeState = {
  selectedNode: null
}

const useNavigationTreeStore = create<TreeNodeState>()(subscribeWithSelector(() => initialState))

export const resetNavigationTreeStore = (): void => {
  useNavigationTreeStore.setState(initialState, true)
}

export const setNavigationTreeSelectedNode = (selectedNode: TreeNode): void => {
  useNavigationTreeStore.setState({
    selectedNode
  })
}

export const useNavigationTreeSelectedNode = (): Maybe<TreeNode> => useNavigationTreeStore(state => state.selectedNode)

```

**Comment by dbritto-dev** [comment]
@viveleroi sure, you can use `subscribe`

**Reply by viveleroi** [reply]
I was under the impression that subscribe was for third party scripts to detect state changes, and was not for use inside a reactive app like React. The problem we had with subscribe before was that every time a react component renders, it subscribes again. We'd have to setup a useEffect for one-time subscribe and then unsub on mount?

**Reply by dbritto-dev** [reply]
@viveleroi yes, you should use it with useEffect
